### ISSUE TITLE
Extract metadata from cmake

### ISSUE THREAD

**Issue by LecrisUT** [scikit-build/scikit-build-core#172]
Currently this project considers the Python build to be the main controller, but many projects, especially when the python api is just a wrapper,  have the cmake as the main build workflow. Therefore we should add more functionalities to have the main information in cmake. This can be achieved by using [`dynamic`](https://peps.python.org/pep-0621/#dynamic) in the `pyproject.toml` and have the builder `scikit-build-core` handle the logic of how to import them. Maybe [this implementation](https://github.com/pypa/flit/blob/97e985f62c62dda99ceca2cf9aab2e25455ec3f1/flit_core/flit_core/buildapi.py#L28-L37) is a good starting point to implement this.

Some metadata that can be set from the [`project()`](https://cmake.org/cmake/help/latest/command/project.html) cmake function:
- version
- name
- description
- url

**Comment by henryiii** [comment]
You cannot set `name` dynamically, this is disallowed by PEP 621. I'm not sure URL would be helpful either, as you have a dict of them and there's no "default" one. Description is supposed to be one line, so duplicating that isn't that big of a deal - this only leaves `version`. That would be very nice to have - see #116. You need CMake to properly compute this, but I _think_ we have `get_requires_for_build_*` in the various places we'd need to produce a version, and we already have parsing tools for the FileAPI of CMake, so we could implement this (and description, and maybe URL if we have something reasonable to map it to).

Meson-Python has a non-PEP 621 mode where you can read all of this info from Meson. I'm not in favor of this, it complicates the backend a lot and makes adding the other metadata much harder (since you can't use PEP 621 fields in that case) - realistic projects need these other fields, like requires-python and such.

**Comment by LecrisUT** [comment]
Would be helpful to also make it expandable. E.g. the project only defines version `1.4.2`, while `-rc1` are only defined in the git via tags. Then we could make it conditional, if there is a release tag, use that as the field, otherwise use the cmake defined. That way we would have feature parity with github repo and pypi uploaded version and also other pip installs do not complain of unsupported versions when we develop locally without git tags.

**Comment by henryiii** [comment]
Sure, links are fine (and encouraged). I think we should be able to add a cmake hook soon as well, I plan to try to write one before 0.3.0.

**Comment by JeanChristopheMorinPerso** [comment]
Hello, I will have a need for this in [MaterialX](https://github.com/AcademySoftwareFoundation/MaterialX). More particularly, the `pyproject.toml` version should be set based on the value in the main `CMakeList.txt` file.

I took a quick look and it seems like this is not actually possible, even with a custom plugin? It would require the "configuration" step to be run before the plugins, which I'm not sure is the case right now. Unless I misunderstand something.

**Comment by JeanChristopheMorinPerso** [comment]
Or is the plan to regex our way out and simply read the CMake file and get the version using a regex?

**Comment by JeanChristopheMorinPerso** [comment]
I ended up writing a plugin:

```python
"""
This is a custom scikit-build-core plugin that will
fetch the MaterialX version from the CMake project.
"""
import os
import tempfile
import subprocess
from pathlib import Path
from typing import FrozenSet, Dict, Optional, Union, List

from scikit_build_core.file_api.query import stateless_query
from scikit_build_core.file_api.reply import load_reply_dir


def dynamic_metadata(
    fields: FrozenSet[str],
    settings: Optional[Dict[str, object]] = None,
) -> Dict[str, Union[str, Dict[str, Optional[str]]]]:
    print("mtx_skbuild_plugin: Computing MaterialX version from CMake...")

    if fields != {"version"}:
        msg = "Only the 'version' field is supported"
        raise ValueError(msg)

    if settings:
        msg = "No inline configuration is supported"
        raise ValueError(msg)

    current_dir = os.path.dirname(__file__)

    with tempfile.TemporaryDirectory() as tmpdir:
        # We will use CMake's file API to get the version
        # instead of parsing the CMakeLists files.

        # First generate the query folder so that CMake can generate replies.
        reply_dir = stateless_query(Path(tmpdir))

        # Run cmake (configure). CMake will generate a reply automatically.
        try:
            subprocess.run(
                [
                    "cmake",
                    "-S",
                    current_dir,
                    "-B",
                    tmpdir,
                    "-DMATERIALX_BUILD_SHARED_LIBS=OFF",
                    "-DMATERIALX_BUILD_PYTHON=OFF",
                    "-DMATERIALX_TEST_RENDER=OFF",
                    "-DMATERIALX_BUILD_TESTS=OFF",
                    "-DMATERIALX_INSTALL_PYTHON=OFF",
                    "-DMATERIALX_BUILD_RENDER=OFF",
                ],
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                check=True,
                text=True,
            )
        except subprocess.CalledProcessError as exc:
            print(exc.stdout)
            raise RuntimeError(
                "Failed to configure project to get the version"
            ) from exc

        # Get the generated replies.
        index = load_reply_dir(reply_dir)

        # Get the version from the CMAKE_PROJECT_VERSION variable.
        entries = [
            entry
            for entry in index.reply.cache_v2.entries
            if entry.name == "CMAKE_PROJECT_VERSION"
        ]

        if not entries:
            raise ValueError("Could not find MaterialX version from CMake project")

        if len(entries) > 1:
            raise ValueError("More than one entry for CMAKE_PROJECT_VERSION found...")

    version = entries[0].value
    print("mtx_skbuild_plugin: Computed version: {0}".format(version))

    return {"version": version}


def get_requires_for_dynamic_metadata(
    _settings: Optional[Dict[str, object]] = None,
) -> List[str]:
    return ["cmake"]
```

**Comment by Freed-Wu** [comment]
If we can set `dynamic = ["version"]`, then get version from `CMakeLists.txt` like `meson-python`, it will be great.

**Comment by henryiii** [comment]
There's a regex plugin built in for this and it's much faster than setting up a CMake run would be.

**Comment by Freed-Wu** [comment]
Another method is write version to CMakeLists.txt, then genereate `__init__.py` to contains `__version__ = "XXX"`. Then

```toml
[project]
dynamic = ["version"]

[tool.setuptools_scm]
```

**Comment by eli-schwartz** [comment]
> There's a regex plugin built in for this and it's much faster than setting up a CMake run would be.

If you're doing a cmake run either way then I can't see how it matters in terms of speed. But the advantage of doing a cmake run is *precisely* that you don't have to deal with regexes.

> Some people, when confronted with a problem, think "I know, I'll use regular expressions." Now they have two problems. 

Regular expressions are almost always a tradeoff of some sort. If it turns out that you can anyways extract the info a better way through something that already exists (a previous cmake run) then really, whyever not?

> Another method is write version to CMakeLists.txt, then genereate `__init__.py` to contains `__version__ = "XXX"`. Then

@Freed-Wu,

I don't really understand this though. setuptools-scm reads version information *from* git or PyPA metadata, and writes *to* an `__init__.py`, and you need a bit more to get scikit-build-core to use setuptools-scm than just adding `tool.setuptools_scm`?

Furthermore, if it requires running cmake to generate your `__init__.py` then we are back to needing to run cmake, at which point it might as well be done properly via the FileAPI with explicit support for only configuring the project once and using it for both metadata and compilation.

**Comment by Freed-Wu** [comment]
> Regular Expression

I heard cmake will generate some json data in `build/` after `cmake -Bbuild`. Maybe read the json can avoid RE.

> Another method

Refer [astyle-wheel](https://github.com/Freed-Wu/astyle-wheel/blob/main/pyproject.toml#L30).

**Comment by henryiii** [comment]
The version-collection CMake run has to be a separate run, since it's collecting metadata information before running CMake properly (which has access to the version!). After you configure is too late (currently). If you are just collecting metadata via the prepare metadata hooks, CMake does not need to run at all if you use the regex.

Long term, it might be possible to cleverly integrate it, and also collect other information too, but for now, the simplest and fastest way is to use the regex plugin. And you can implement it as a custom plugin.

**Comment by alexreinking** [comment]
I'm currently in a situation where I want to package multiple versions of a third-party C++ dependency (built in a manylinux container). The workflow I had envisioned was:

1. Write a wrapper CMakeLists.txt that accepts a git tag and then uses FetchContent to import the build.
2. Set the Python package version from the sub-project's version variable.
3. Write install rules that scikit-build-core will like.

Clearly, regex cannot be used to do (2) as the third-party sources don't exist yet.

What should I do instead?

**Comment by alexreinking** [comment]
> I could use an example to properly visualize this. But why not invert the source of truth, i.e. have the `pyproject.toml`(s) be static and on the CMake side have a switch around `SKBUILD_PROJECT_NAME` to do the appropriate FetchContent with the version from `SKBUILD_PROJECT_VERSION_FULL`.
> 
> But then what parameters do you pass to FetchContent? Couldn't you centralize that source to be readable by both?

Sure -- the specific thing I'm trying to do is build multiple versions of LLVM, plus the state of the main branch at build time.

My FetchContent call looks like this:

```
set(GIT_TAG "main"
    CACHE STRING "LLVM reference to package")

if (GIT_TAG STREQUAL "main")
    set(url "heads/main.tar.gz")
elseif (GIT_TAG MATCHES "^llvmorg-[0-9]+\\.[0-9]+\\.[0-9]+$")
    set(url "tags/${GIT_TAG}.tar.gz")
else ()
    message(FATAL_ERROR "GIT_TAG should be set to a release tag or `main`")
endif ()

include(FetchContent)
FetchContent_Declare(
    LLVM
    URL "https://github.com/llvm/llvm-project/archive/refs/${url}"
    SOURCE_SUBDIR llvm
)
FetchContent_MakeAvailable(LLVM)
```

This way, I can set `-Ccmake.define.GIT_TAG=llvmorg-18.1.8`, say.

If I were to invert the source of truth, I could easily encode the tag as `llvmorg-${SKBUILD_PROJECT_VERSION}`. However, the complicating factor is that I need nightlies. I don't know what the version should be until I pull / download.

**Comment by LecrisUT** [comment]
> If I were to invert the source of truth, I could easily encode the tag as `llvmorg-${SKBUILD_PROJECT_VERSION}`. However, the complicating factor is that I need nightlies. I don't know what the version should be until I pull / download.

Ah ok I see that. Supporting nightlies is an intriguing challenge. I wonder if `FetchContent` works in script mode, if so you could write a small cmake wrapper and extract the version from there. The previous https://github.com/scikit-build/scikit-build-core/issues/172#issuecomment-1575689576 is a nice starting point for that. You would need the cache variables which might be a bit tricky to pull, but if you can use the environment variable instead you could have all the pieces you need now.

---

> > pyproject.toml(s)
> 
> Is there a way to have multiple such files without essentially duplicating them over and over?

In principle yes. What you would do is create subfolders for each subproject with a pyproject.toml in it and either reference the parent folder's CMakeLists or probably more robust, have a small CMakeLists that does that.

The complication is accounting for `sdist` builds (because by default you don't have access to the parent files), in which case you should `sdist.include` and account for the different file structure in each build. This last part might not work currently, but there are some revisions being done around the related architecture for the next minor release.

**Comment by alexreinking** [comment]
> Ah ok I see that. Supporting nightlies is an intriguing challenge. I wonder if `FetchContent` works in script mode, if so you could write a small cmake wrapper and extract the version from there. The previous [#172 (comment)](https://github.com/scikit-build/scikit-build-core/issues/172#issuecomment-1575689576) is a nice starting point for that. You would need the cache variables which might be a bit tricky to pull, but if you can use the environment variable instead you could have all the pieces you need now.

I would be more interested in configuring the build in-place twice. Run it the first time with the `SKBUILD_STATE` set to `metadata_cmake`[^1] or the like so the build knows what's happening (and can even exit early!). Then configure a second time with all the resolved metadata.

The benefit of doing it in-place is that the results of `FetchContent` calls would be cached. Some dependencies, like LLVM, tensorflow, pytorch, etc. take a long time even to clone (several minutes). I'm already setting `build-dir = "build/{wheel_tag}"` in my config.

This behavior would absolutely need to be opt-in, of course.

[^1]: By the way, what is `metadata_wheel` used for?

> In principle yes. What you would do is create subfolders for each subproject with a pyproject.toml in it and either reference the parent folder's CMakeLists or probably more robust, have a small CMakeLists that does that.

By "each subproject", you mean parallel major versions (plus main) of LLVM? I'm not 100% sure I see what you're saying here.

> The complication is accounting for `sdist` builds (because by default you don't have access to the parent files), in which case you should `sdist.include` and account for the different file structure in each build. This last part might not work currently, but there are some revisions being done around the related architecture for the next minor release.

For my purposes, I'm not too concerned about sdist builds. The plan is for all this to run in a relatively controlled setting, not on the public PyPI instance (the LLVM binaries are too large, anyway). I'd rather hear about a missing wheel than have a runner unexpectedly spend four hours building LLVM.

**Comment by LecrisUT** [comment]
> I would be more interested in configuring the build in-place twice.

That is also possible but you need to read the `build-dir` manually and also set it statically in the `pyproject.toml` (see the interface you have available in the comment example `dynamic_metadata` and `get_requires_for_dynamic_metadata`).

About the `SKBUILD_STATE` I don't think I've checked that logic, but I believe the cmake configure is done only once currently for each build call.

---

> By "each subproject", do you mean several parallel major versions (plus main) of LLVM? I'm not 100% sure I see what you're saying here.

Not the LLVM example specifically, just more generally. Anything that would have a different `pyproject.toml` (specifically a different `project.name`) would be a different (sub)project.

For your case it's the same project so it doesn't really make sense to do such separation, but if you build each components separately, it may be a potential design.

---

Btw, there are quite a few people working on llvm/clang building and packaging, have you got in touch with some of them? https://github.com/llvm/llvm-project/pull/125806 was the most recent one.
